core/contracts/Registry.sol

```diff
@@ -57,58 +55,84 @@ contract Registry is RegistryInterface, MultiRole {
+        derivative.valid = DerivativeValidity.Valid;
         for (uint i = 0; i < parties.length; i = i.add(1)) {
-            partiesMap.parties[parties[i]] = true;
+            derivative.parties[parties[i]] = true;
+            partyMembersToDerivatives[parties[i]].push(derivativeAddress);
         }
 
-        address[] memory partiesForEvent = parties;
-        emit NewDerivativeRegistered(derivativeAddress, msg.sender, partiesForEvent);
+        emit NewDerivativeRegistered(derivativeAddress, msg.sender, parties);
     }
 
-    function isDerivativeRegistered(address derivative) external view returns (bool isRegistered) {
-        return derivativePointers[derivative].valid == PointerValidity.Valid;
+    function isDerivativeRegistered(address derivative) external view returns (bool) {
+        return addressToDerivatives[derivative].valid == DerivativeValidity.Valid;
     }
 
-    function getRegisteredDerivatives(address party) external view returns (address[] memory derivatives) {
-        // This is not ideal - we must statically allocate memory arrays. To be safe, we make a temporary array as long
-        // as registeredDerivatives. We populate it with any derivatives that involve the provided party. Then, we copy
-        // the array over to the return array, which is allocated using the correct size. Note: this is done by double
-        // copying each value rather than storing some referential info (like indices) in memory to reduce the number
-        // of storage reads. This is because storage reads are far more expensive than extra memory space (~100:1).
-        address[] memory tmpDerivativeArray = new address[](registeredDerivatives.length);
-        uint outputIndex = 0;
-        for (uint i = 0; i < registeredDerivatives.length; i = i.add(1)) {
-            address derivative = registeredDerivatives[i];
-            if (derivativesToParties[derivative].parties[party]) {
-                // Copy selected derivative to the temporary array.
-                tmpDerivativeArray[outputIndex] = derivative;
-                outputIndex = outputIndex.add(1);
+    function getRegisteredDerivatives(address party) external view returns (address[] memory) {
+        return partyMembersToDerivatives[party];
+    }
+
+    function addPartyToDerivative(address party) external {
+        address derivativeAddress = msg.sender;
+
+        require(
+            addressToDerivatives[derivativeAddress].valid == DerivativeValidity.Valid,
+            "Can add to valid derivative"
+        );
+        require(addressToDerivatives[derivativeAddress].parties[party] == false, "Can only register a party once");
+
+        addressToDerivatives[derivativeAddress].parties[party] = true;
+        partyMembersToDerivatives[party].push(derivativeAddress);
+
+        emit PartyMemberAdded(derivativeAddress, party);
+    }
+
+    function removedPartyFromDerivative(address party) external {
```

Comment:
nit: why `removed` rather than `remove`?

Reply:
will fix this.